File: sdks/php/tests/HandlerTest.php
<?php

namespace Autonoma\Sdk\Tests;

use Autonoma\Sdk\Handler;
use Autonoma\Sdk\Hmac;
use Autonoma\Sdk\Scenario;
use Autonoma\Sdk\Types\HandlerConfig;
use Autonoma\Sdk\Types\HandlerRequest;
use Autonoma\Sdk\Types\ScenarioDefinition;
use Autonoma\Sdk\Types\ScenarioDownContext;
use Autonoma\Sdk\Types\ScenarioUpContext;
use Autonoma\Sdk\Types\ScenarioUpResult;
use PHPUnit\Framework\TestCase;

class HandlerTest extends TestCase
{
    private function standardScenario(?callable $down = null): ScenarioDefinition
    {
        return Scenario::defineScenario(
            name: 'standard',
            description: 'A standard seeded environment',
            up: fn(ScenarioUpContext $ctx) => new ScenarioUpResult(
                auth: ['headers' => ['Authorization' => 'Bearer ' . $ctx->testRunId]],
                teardown: ['userId' => 'user-' . $ctx->testRunId],
            ),
            down: $down,
        );
    }

    private function emptyScenario(): ScenarioDefinition
    {
        return Scenario::defineScenario(
            name: 'empty',
            description: 'Nothing seeded',
            up: fn(ScenarioUpContext $ctx) => new ScenarioUpResult(),
        );
    }

    /** @param ScenarioDefinition[]|null $scenarios */
    private function makeConfig(?array $scenarios = null): HandlerConfig
    {
        return new HandlerConfig(
            sharedSecret: 'shared',
            signingSecret: 'signing',
            scenarios: $scenarios ?? [$this->standardScenario(), $this->emptyScenario()],
        );
    }

    private function makeRequest(string $body, string $secret = 'shared'): HandlerRequest
    {
        return new HandlerRequest(
            body: $body,
            headers: ['x-signature' => Hmac::signBody($body, $secret)],
        );
    }

    // --- request gate ---

    public function testRejectsInvalidSignature(): void
    {
        $req = new HandlerRequest(body: '{"action":"discover"}', headers: ['x-signature' => 'invalid']);
        $res = Handler::handleRequest($this->makeConfig(), $req);
        $this->assertSame(401, $res->status);
        $this->assertSame('INVALID_SIGNATURE', $res->body['code']);
    }

    public function testRejectsSameSecrets(): void
    {
        $config = new HandlerConfig(sharedSecret: 'same', signingSecret: 'same');
        $res = Handler::handleRequest($config, $this->makeRequest('{"action":"discover"}', 'same'));
        $this->assertSame(500, $res->status);
        $this->assertSame('SAME_SECRETS', $res->body['code']);
    }

    public function testRejectsInvalidJson(): void
    {
        $res = Handler::handleRequest($this->makeConfig(), $this->makeRequest('not json'));
        $this->assertSame(400, $res->status);
        $this->assertSame('INVALID_BODY', $res->body['code']);
    }

    public function testRejectsMissingAction(): void
    {
        $res = Handler::handleRequest($this->makeConfig(), $this->makeRequest('{}'));
        $this->assertSame(400, $res->status);
        $this->assertSame('INVALID_BODY', $res->body['code']);
    }

    public function testRejectsUnknownAction(): void
    {
        $res = Handler::handleRequest($this->makeConfig(), $this->makeRequest('{"action":"nonexistent"}'));
        $this->assertSame(400, $res->status);
        $this->assertSame('UNKNOWN_ACTION', $res->body['code']);
    }

    // --- discover ---

    public function testDiscoverListsScenarios(): void
    {
        $res = Handler::handleRequest($this->makeConfig(), $this->makeRequest('{"action":"discover"}'));
        $this->assertSame(200, $res->status);
        $this->assertSame('2.0', $res->body['version']);
        $this->assertCount(2, $res->body['scenarios']);
        $this->assertSame('standard', $res->body['scenarios'][0]['name']);
        $this->assertNotEmpty($res->body['scenarios'][0]['description']);
        // discover must never leak a create/schema shape in v2.
        $this->assertArrayNotHasKey('schema', $res->body);
    }

    // --- up ---

    public function testUpReturnsEnvelope(): void
    {
        $body = json_encode(['action' => 'up', 'scenario' => ['name' => 'standard'], 'testRunId' => 'run-1']);
        $res = Handler::handleRequest($this->makeConfig(), $this->makeRequest($body));

        $this->assertSame(200, $res->status);
        $this->assertSame('2.0', $res->body['version']);
        $this->assertCount(3, explode('.', $res->body['teardownToken']));
        $this->assertSame(3600, $res->body['expiresInSeconds']);
        // The duplicated plaintext refs and the old refsToken field are gone.
        $this->assertArrayNotHasKey('refs', $res->body);
        $this->assertArrayNotHasKey('refsToken', $res->body);
        $this->assertSame('Bearer run-1', $res->body['auth']['headers']['Authorization']);
    }

    public function testUpCustomExpires(): void
    {
        $config = new HandlerConfig(
            sharedSecret: 'shared',
            signingSecret: 'signing',
            scenarios: [$this->standardScenario(), $this->emptyScenario()],
            expiresInSeconds: 60,
        );
        $body = json_encode(['action' => 'up', 'scenario' => ['name' => 'empty'], 'testRunId' => 'r']);
        $res = Handler::handleRequest($config, $this->makeRequest($body));

        $this->assertSame(60, $res->body['expiresInSeconds']);
        // The empty scenario returns nothing, so no auth on the envelope.
        $this->assertArrayNotHasKey('auth', $res->body);
    }

    public function testUpUnknownEnvironment(): void
    {
        $body = json_encode(['action' => 'up', 'scenario' => ['name' => 'does-not-exist'], 'testRunId' => 'r']);
        $res = Handler::handleRequest($this->makeConfig(), $this->makeRequest($body));
        $this->assertSame(400, $res->status);
        $this->assertSame('UNKNOWN_ENVIRONMENT', $res->body['code']);
    }

    public function testUpMissingScenarioName(): void
    {
        $body = json_encode(['action' => 'up', 'testRunId' => 'r']);
        $res = Handler::handleRequest($this->makeConfig(), $this->makeRequest($body));
        $this->assertSame(400, $res->status);
        $this->assertSame('INVALID_BODY', $res->body['code']);
    }

    public function testUpAcceptsPlainArrayResult(): void
    {
        $arrayScenario = Scenario::defineScenario(
            name: 'plain',
            description: 'up returns a plain array',
            up: fn(ScenarioUpContext $ctx) => [
                'auth' => ['headers' => ['X-Token' => 'plain']],
            ],
        );
        $body = json_encode(['action' => 'up', 'scenario' => ['name' => 'plain'], 'testRunId' => 'r']);
        $res = Handler::handleRequest($this->makeConfig([$arrayScenario]), $this->makeRequest($body));
        $this->assertSame(200, $res->status);
        $this->assertSame('plain', $res->body['auth']['headers']['X-Token']);
    }

    // --- down ---

    public function testDownValidToken(): void
    {
        $downCalls = [];
        $down = function (ScenarioDownContext $ctx) use (&$downCalls): void {
            $downCalls[] = $ctx->name . ':' . $ctx->testRunId;
        };
        $config = $this->makeConfig([$this->standardScenario($down), $this->emptyScenario()]);

        $upBody = json_encode(['action' => 'up', 'scenario' => ['name' => 'standard'], 'testRunId' => 'run-td']);
        $upRes = Handler::handleRequest($config, $this->makeRequest($upBody));
        $token = $upRes->body['teardownToken'];

        $downBody = json_encode(['action' => 'down', 'teardownToken' => $token, 'testRunId' => 'run-td']);
        $downRes = Handler::handleRequest($config, $this->makeRequest($downBody));

        $this->assertSame(200, $downRes->status);
        $this->assertTrue($downRes->body['ok']);
        $this->assertSame(['standard:run-td'], $downCalls);
    }

    public function testDownRoutesByTokenEnvironment(): void
    {
        $downCalls = [];
        $down = function (ScenarioDownContext $ctx) use (&$downCalls): void {
            $downCalls[] = $ctx->name . ':' . $ctx->testRunId;
        };
        $config = $this->makeConfig([$this->standardScenario($down), $this->emptyScenario()]);

        $upBody = json_encode(['action' => 'up', 'scenario' => ['name' => 'standard'], 'testRunId' => 'run-tok']);
        $token = Handler::handleRequest($config, $this->makeRequest($upBody))->body['teardownToken'];

        // No scenario.name on the down request - the handler must recover it from
        // the verified token's environment.
        $downBody = json_encode(['action' => 'down', 'teardownToken' => $token]);
        $downRes = Handler::handleRequest($config, $this->makeRequest($downBody));

        $this->assertSame(200, $downRes->status);
        $this->assertSame(['standard:run-tok'], $downCalls);
    }

    public function testDownInvalidTeardownToken(): void
    {
        $body = json_encode(['action' => 'down', 'teardownToken' => 'tampered.token.value']);
        $res = Handler::handleRequest($this->makeConfig(), $this->makeRequest($body));
        $this->assertSame(403, $res->status);
        $this->assertSame('INVALID_TEARDOWN_TOKEN', $res->body['code']);
    }

    public function testDownMissingTeardownToken(): void
    {
        $res = Handler::handleRequest($this->makeConfig(), $this->makeRequest('{"action":"down"}'));
        $this->assertSame(400, $res->status);
        $this->assertSame('INVALID_BODY', $res->body['code']);
    }

    public function testEndpointAlwaysEnabled(): void
    {
        // allowProduction is a deprecated no-op: discover serves regardless.
        $config = new HandlerConfig(
            sharedSecret: 'shared',
            signingSecret: 'signing',
            scenarios: [$this->standardScenario()],
            allowProduction: false,
        );
        $res = Handler::handleRequest($config, $this->makeRequest('{"action":"discover"}'));
        $this->assertSame(200, $res->status);
    }
}

File: sdks/rust/tests/handler_test.rs
//! Handler integration tests for the Scenario v2 SDK.

use autonoma_sdk::errors::AutonomaError;
use autonoma_sdk::handler::handle_request;
use autonoma_sdk::hmac::sign_body;
use autonoma_sdk::scenario::{define_scenario, define_scenario_up_only, Scenario};
use autonoma_sdk::types::{
    AuthResult, HandlerConfig, HandlerRequest, ScenarioDownContext, ScenarioUpContext,
    ScenarioUpResult,
};
use serde_json::{json, Value};
use std::collections::HashMap;
use std::future::Future;
use std::pin::Pin;
use std::sync::{Arc, Mutex};

// --- Scenarios ---

fn standard_up(
    ctx: &ScenarioUpContext,
) -> Pin<Box<dyn Future<Output = Result<ScenarioUpResult, AutonomaError>> + Send + '_>> {
    let test_run_id = ctx.test_run_id.clone();
    Box::pin(async move {
        let mut headers = HashMap::new();
        headers.insert(
            "Authorization".to_string(),
            format!("Bearer {}", test_run_id),
        );
        Ok(ScenarioUpResult {
            auth: Some(AuthResult {
                headers: Some(headers),
                ..Default::default()
            }),
            teardown: Some(json!({ "userId": format!("user-{}", test_run_id) })),
        })
    })
}

fn recording_down(
    calls: Arc<Mutex<Vec<String>>>,
) -> impl for<'a> Fn(
    &'a ScenarioDownContext,
) -> Pin<Box<dyn Future<Output = Result<(), AutonomaError>> + Send + 'a>>
       + Send
       + Sync
       + 'static {
    move |ctx: &ScenarioDownContext| {
        let entry = format!("{}:{}", ctx.name, ctx.test_run_id);
        let calls = calls.clone();
        Box::pin(async move {
            calls.lock().unwrap().push(entry);
            Ok(())
        })
    }
}

fn test_scenarios(down_calls: Option<Arc<Mutex<Vec<String>>>>) -> Vec<Box<dyn Scenario>> {
    let standard = match down_calls {
        Some(calls) => define_scenario(
            "standard",
            "A standard seeded environment",
            standard_up,
            Some(recording_down(calls)),
        ),
        None => define_scenario_up_only("standard", "A standard seeded environment", standard_up),
    };
    vec![
        standard,
        define_scenario_up_only("empty", "Nothing seeded", |_ctx| {
            Box::pin(async move { Ok(ScenarioUpResult::default()) })
        }),
    ]
}

#[allow(deprecated)]
fn base_config(down_calls: Option<Arc<Mutex<Vec<String>>>>) -> HandlerConfig {
    HandlerConfig {
        shared_secret: "shared".to_string(),
        signing_secret: "signing".to_string(),
        scenarios: test_scenarios(down_calls),
        expires_in_seconds: None,
        allow_production: false,
        sdk: None,
    }
}

fn signed_req(body: &Value, secret: &str) -> HandlerRequest {
    signed_req_raw(&body.to_string(), secret)
}

fn signed_req_raw(body: &str, secret: &str) -> HandlerRequest {
    let mut headers = HashMap::new();
    headers.insert("x-signature".to_string(), sign_body(body, secret));
    HandlerRequest {
        body: body.to_string(),
        headers,
    }
}

// --- Request gate ---

#[tokio::test]
async fn rejects_invalid_signature() {
    let req = HandlerRequest {
        body: r#"{"action":"discover"}"#.to_string(),
        headers: HashMap::from([("x-signature".to_string(), "invalid".to_string())]),
    };
    let resp = handle_request(&base_config(None), &req).await;
    assert_eq!(resp.status, 401);
    assert_eq!(resp.body["code"], "INVALID_SIGNATURE");
}

#[tokio::test]
async fn rejects_same_secrets() {
    #[allow(deprecated)]
    let config = HandlerConfig {
        shared_secret: "same".to_string(),
        signing_secret: "same".to_string(),
        scenarios: vec![],
        expires_in_seconds: None,
        allow_production: false,
        sdk: None,
    };
    let req = HandlerRequest {
        body: r#"{"action":"discover"}"#.to_string(),
        headers: HashMap::from([("x-signature".to_string(), "x".to_string())]),
    };
    let resp = handle_request(&config, &req).await;
    assert_eq!(resp.status, 500);
    assert_eq!(resp.body["code"], "SAME_SECRETS");
}

#[tokio::test]
async fn rejects_invalid_body() {
    let resp = handle_request(&base_config(None), &signed_req_raw("not json", "shared")).await;
    assert_eq!(resp.status, 400);
    assert_eq!(resp.body["code"], "INVALID_BODY");
}

#[tokio::test]
async fn rejects_unknown_action() {
    let req = signed_req(&json!({ "action": "nonexistent" }), "shared");
    let resp = handle_request(&base_config(None), &req).await;
    assert_eq!(resp.status, 400);
    assert_eq!(resp.body["code"], "UNKNOWN_ACTION");
}

// --- discover ---

#[tokio::test]
async fn discover_lists_scenarios() {
    let req = signed_req(&json!({ "action": "discover" }), "shared");
    let resp = handle_request(&base_config(None), &req).await;
    assert_eq!(resp.status, 200);
    assert_eq!(resp.body["version"], "2.0");

    let scenarios = resp.body["scenarios"].as_array().unwrap();
    assert_eq!(scenarios.len(), 2);
    assert_eq!(scenarios[0]["name"], "standard");
    assert!(!scenarios[0]["description"].as_str().unwrap().is_empty());

    // discover must never leak a create/schema shape in v2.
    assert!(resp.body.get("schema").is_none());
}

// --- up ---

#[tokio::test]
async fn up_returns_envelope() {
    let body = json!({ "action": "up", "scenario": { "name": "standard" }, "testRunId": "run-1" });
    let resp = handle_request(&base_config(None), &signed_req(&body, "shared")).await;
    assert_eq!(resp.status, 200);
    assert_eq!(resp.body["version"], "2.0");

    let token = resp.body["teardownToken"].as_str().unwrap();
    assert_eq!(token.split('.').count(), 3);
    assert_eq!(resp.body["expiresInSeconds"], 3600);
    assert_eq!(resp.body["auth"]["headers"]["Authorization"], "Bearer run-1");
    // The duplicated plaintext refs and the old refsToken field are gone.
    assert!(resp.body.get("refs").is_none());
    assert!(resp.body.get("refsToken").is_none());
}

#[tokio::test]
async fn up_custom_expires() {
    let mut config = base_config(None);
    config.expires_in_seconds = Some(60);
    let body = json!({ "action": "up", "scenario": { "name": "empty" }, "testRunId": "r" });
    let resp = handle_request(&config, &signed_req(&body, "shared")).await;
    assert_eq!(resp.body["expiresInSeconds"], 60);
    // The empty scenario returns nothing, so no auth on the envelope.
    assert!(resp.body.get("auth").is_none());
}

#[tokio::test]
async fn up_unknown_environment() {
    let body = json!({ "action": "up", "scenario": { "name": "does-not-exist" }, "testRunId": "r" });
    let resp = handle_request(&base_config(None), &signed_req(&body, "shared")).await;
    assert_eq!(resp.status, 400);
    assert_eq!(resp.body["code"], "UNKNOWN_ENVIRONMENT");
}

#[tokio::test]
async fn up_missing_scenario_name() {
    let body = json!({ "action": "up", "testRunId": "r" });
    let resp = handle_request(&base_config(None), &signed_req(&body, "shared")).await;
    assert_eq!(resp.status, 400);
    assert_eq!(resp.body["code"], "INVALID_BODY");
}


// --- down ---

#[tokio::test]
async fn down_valid_token() {
    let down_calls = Arc::new(Mutex::new(Vec::new()));
    let config = base_config(Some(down_calls.clone()));

    let up_body = json!({ "action": "up", "scenario": { "name": "standard" }, "testRunId": "run-td" });
    let up_resp = handle_request(&config, &signed_req(&up_body, "shared")).await;
    let token = up_resp.body["teardownToken"].as_str().unwrap().to_string();

    let down_body = json!({ "action": "down", "teardownToken": token, "testRunId": "run-td" });
    let down_resp = handle_request(&config, &signed_req(&down_body, "shared")).await;
    assert_eq!(down_resp.status, 200);
    assert_eq!(down_resp.body["ok"], true);

    let calls = down_calls.lock().unwrap();
    assert_eq!(calls.len(), 1);
    assert_eq!(calls[0], "standard:run-td");
}

#[tokio::test]
async fn down_routes_by_token_environment() {
    let down_calls = Arc::new(Mutex::new(Vec::new()));
    let config = base_config(Some(down_calls.clone()));

    let up_body = json!({ "action": "up", "scenario": { "name": "standard" }, "testRunId": "run-tok" });
    let token = handle_request(&config, &signed_req(&up_body, "shared"))
        .await
        .body["teardownToken"]
        .as_str()
        .unwrap()
        .to_string();

    // No scenario.name on the down request - the handler must recover it from
    // the verified token's environment.
    let down_body = json!({ "action": "down", "teardownToken": token });
    let down_resp = handle_request(&config, &signed_req(&down_body, "shared")).await;
    assert_eq!(down_resp.status, 200);

    let calls = down_calls.lock().unwrap();
    assert_eq!(calls.len(), 1);
    assert_eq!(calls[0], "standard:run-tok");
}

#[tokio::test]
async fn down_invalid_teardown_token() {
    let body = json!({ "action": "down", "teardownToken": "tampered.token.value" });
    let resp = handle_request(&base_config(None), &signed_req(&body, "shared")).await;
    assert_eq!(resp.status, 403);
    assert_eq!(resp.body["code"], "INVALID_TEARDOWN_TOKEN");
}

#[tokio::test]
async fn down_missing_teardown_token() {
    let body = json!({ "action": "down" });
    let resp = handle_request(&base_config(None), &signed_req(&body, "shared")).await;
    assert_eq!(resp.status, 400);
    assert_eq!(resp.body["code"], "INVALID_BODY");
}

#[tokio::test]
async fn endpoint_always_enabled() {
    // allow_production is a deprecated no-op: discover serves regardless.
    let req = signed_req(&json!({ "action": "discover" }), "shared");
    let resp = handle_request(&base_config(None), &req).await;
    assert_eq!(resp.status, 200);
}
